tests/mocha/field_date_test.js

```diff
@@ -0,0 +1,176 @@
+      assertValueDefault(dateField);
+    });
+    test('Undefined', function() {
+      var dateField = new Blockly.FieldDate(undefined);
+      assertValueDefault(dateField);
+    });
+    test('Non-Parsable String', function() {
+      var dateField = new Blockly.FieldDate('bad');
+      assertValueDefault(dateField);
+    });
+    test('2020-02-20', function() {
+      var dateField = new Blockly.FieldDate('2020-02-20');
+      assertValue(dateField, '2020-02-20');
+    });
+    test('Invalid Date - Month(2020-13-20)', function() {
+      // TODO: What do we want it to do in this case?
+      var dateField = new Blockly.FieldDate('2020-13-20');
+      assertValueDefault(dateField);
+    });
+    test('Invalid Date - Day(2020-02-32)', function() {
+      // TODO: What do we want it to do in this case?
+      var dateField = new Blockly.FieldDate('2020-02-32');
+      assertValueDefault(dateField);
+    });
+  });
+  suite('fromJson', function() {
+    test('Empty', function() {
+      var dateField = Blockly.FieldDate.fromJson({});
+      assertValueDefault(dateField);
+    });
+    test('Null', function() {
+      var dateField = Blockly.FieldDate.fromJson({ date: null });
+      assertValueDefault(dateField);
+    });
+    test('Undefined', function() {
+      var dateField = Blockly.FieldDate.fromJson({ date: undefined });
+      assertValueDefault(dateField);
+    });
+    test('Non-Parsable String', function() {
+      var dateField = Blockly.FieldDate.fromJson({ date: 'bad' });
+      assertValueDefault(dateField);
+    });
+    test('2020-02-20', function() {
+      var dateField = Blockly.FieldDate.fromJson({ date: '2020-02-20' });
+      assertValue(dateField, '2020-02-20');
+    });
+    test('Invalid Date - Month(2020-13-20)', function() {
+      // TODO: What do we want it to do in this case?
+      var dateField = Blockly.FieldDate.fromJson({ date: '2020-13-20' });
+      assertValueDefault(dateField);
+    });
+    test('Invalid Date - Day(2020-02-32)', function() {
+      // TODO: What do we want it to do in this case?
+      var dateField = Blockly.FieldDate.fromJson({ date: '2020-02-32' });
+      assertValueDefault(dateField);
+    });
+  });
+  suite('setValue', function() {
+    test('Empty -> Null', function() {
+      var dateField = new Blockly.FieldDate();
+      dateField.setValue(null);
+      assertValueDefault(dateField);
+    });
+    test('Value -> Null', function() {
+      var dateField = new Blockly.FieldDate('2020-02-20');
+      dateField.setValue(null);
+      assertValue(dateField, '2020-02-20');
+    });
+    test('Empty -> Undefined', function() {
+      var dateField = new Blockly.FieldDate();
+      dateField.setValue(undefined);
+      assertValueDefault(dateField);
+    });
+    test('Value -> Undefined', function() {
+      var dateField = new Blockly.FieldDate('2020-02-20');
+      dateField.setValue(undefined);
+      // TODO: Is this how we want this to work?
```

Comment:
As above, probably want it to leave the current value instead of reverting to the default.